README.md

```diff
@@ -1,2 +1,104 @@
+
+MySQL designed for container-native deployment on Joyent's Triton platform. This repo serves as a blueprint for automatically setting up replication, backups, and failover without human intervention.
+
+---
+
+## Architecture
+
+A running cluster includes the following components:
+- [Consul](https://www.consul.io/): used to coordinate replication and failover
+- [MySQL](https://www.mysql.com/): we're using MySQL5.6 via [Percona Server](https://www.percona.com/software/mysql-database/percona-server), and [`xtrabackup`](https://www.percona.com/software/mysql-database/percona-xtrabackup) for running hot snapshots.
+- [Manta](https://www.joyent.com/object-storage): for securely and durably storing our MySQL snapshots.
+- [Containerbuddy](http://containerbuddy.io): included in our MySQL containers orchestrate bootstrap behavior and coordinate replication using keys and checks stored in Consul in the `onStart`, `health`, and `onChange` handlers.
+- `triton-mysql.py`: a small Python application that Containerbuddy will call into to do the heavy lifting of bootstrapping MySQL.
+
+When a new MySQL node is started, Containerbuddy's `onStart` handler will call into `triton-mysql.py`.
+
+
+### Bootstrapping via `onStart` handler
+
+The first thing the `triton-mysql.py` application does is to ask Consul whether a primary node exists. If not, the application will atomically mark the node as primary in Consul and then bootstrap the node as a new primary. Bootstrapping a primary involves setting up users (root, default, and replication), and creating a default schema. Once the primary bootstrap process is complete, it will use `xtrabackup` to create a backup and upload it to Manta. The application then writes a TTL key to Consul which will tell us when next to run a backup, and a non-expiring key that records the path on Manta where the most recent backup was stored.
+
+If a primary already exists, then the application will ask Consul for the path to the most recent backup and download it. The application will then ask Consul for the IP address of the primary and set up replication from that primary before allowing the new replica to join the cluster.
+
+### Maintenance via `health` handler
+
+The Containerbuddy `health` handler calls `triton-mysql.py health`. If the node is a replica, it merely checks the health of the instance by making a `SELECT 1` against the database and exiting with exit code 0 (which Containerbuddy then interprets as a message to send a heartbeat to Consul).
+
+If the node is primary, the handler will ask Consul if the TTL key for backups have expired or whether the binlog has been rotated. If either case is true, the application will create a new snapshot, upload it to Manta, and write the appropriate keys to Consul to tell replicas where to find the backup.
+
+### Failover via `onChange` handler
+
+*(Note: this is all TODO and the lock semantic won't quite work like this the way the code is now)*
```

Comment:
This is going to be a little tricky. There's a health heartbeat for `mysql-primary` already, but that API doesn't have a locking semantic. I think we'll need to use Consul's sessions API to make this work -- those can have both locks and TTLs.